### DISCUSSION TITLE
Windows App Manifest

### DISCUSSION THREAD

**Discussion by AdarshSrikanth** [marcelotduarte/cx_Freeze#997] [Q&A] [answered]
I'm trying to set compatibility option to the executable built with Cx-Freeze. I'm trying to do this with the help of a windows application manifest file.

1. Is there a way to bundle the windows application manifest file to an exe built with Cx-Freeze.
2. If not, is there any other way to specify the compatibility of the application built with Cx-Freeze.

**Comment by marcelotduarte** [comment]
Are you talking about what's on this [link](https://docs.microsoft.com/en-us/windows/win32/sbscs/application-manifests)?

I am not very aware of this, but I have read a little, and apparently, there is a flaw in the documentation, that this term 'assembly' means that it only works in .NET app, not in common C app, as is the case with the cx-freeze loader.

If I'm wrong, then just add an XML file in the same directory as the .exe.

**Reply by AdarshSrikanth** [reply]
Thanks marcel

The link you pointed is the same MSDN section I'm referring to as well. The documentation says it is is applicable for all Win32 applications. So this is not specific to .NET. To verify this I tried applying an app manifest to a CPP application, and it worked as expected. 

Under the same section in MSDN I found this Win SDK tool [mt.exe](https://docs.microsoft.com/en-us/windows/win32/sbscs/mt-exe).

I tired this tool on my exe file generated with Cx-Freeze with a sample manifest file. With this I was able to add the app manifest to my executable file.

Packing this manifest is important to us because, without setting app compatibility, Windows treats the application to be running  with Windows 8 (major:8 minor:2 build:9200) compatibility on accessing deprecated win32 apis.  

Even latest Python (3.9.2) libraries use some of those deprecated win32 apis.

For example:
sys.getwindowsversion() give (major=10, minor=0, build=18362, platform=2, service_pack='') when interpreted or ran locally.
but if the same call is part of an executable without a manifest, it returns (major=6, minor=2, build=9200, platform=2, service_pack='')

Python docs tells us that getWindowsVersion() is a wrapper over win32 api - GetVersion() which is deprecated since Windows 8

Adding my References:
1. https://docs.python.org/3.9/library/sys.html?highlight=getwindowsversion#sys.getwindowsversion
2. https://docs.microsoft.com/en-us/windows/win32/api/sysinfoapi/nf-sysinfoapi-getversionexa

**Reply by AdarshSrikanth** [reply]
is it possible to support this assembly referencing through manifest files by Cx-freeze?

**Reply by marcelotduarte** [reply]
There are a manifest at https://github.com/marcelotduarte/cx_Freeze/tree/main/source/bases
But it appears to be incorrect/unued ... 
If you can make a PR, I'll accept it.

**Reply by AdarshSrikanth** [reply]
Will do. Thanks 🙂

**Reply by marcelotduarte** [reply]
Correcting a piece of information: I observed that the manifest in the code base is used only when compiled using GCC (MinGW).

**Reply by AdarshSrikanth** [reply]
As it is possible that different manifests can be supplied to multiple executables, I was thinking of an option to pass on the manifest file path as a parameter to "Executable( )" and apply the manifest to required exe files. That way we can apply user supplied manifest to their executable

**Reply by marcelotduarte** [reply]
@AdarshSrikanth  I'm thinking about looking at this problem. If you have made any code, let me know.

**Reply by AdarshSrikanth** [reply]
No, I have not worked on any code for this. I have been using the mt.exe tool to apply my manifest to the executables built with cx-freeze

**Reply by marcelotduarte** [reply]
[cx_Freeze 6.10](https://github.com/marcelotduarte/cx_Freeze/releases/tag/6.10) has just been released.

**Comment by marcelotduarte** [comment, marked as answer]
cx_Freeze 6.10+ accepts the manifest and uac-admin options:
https://cx-freeze.readthedocs.io/en/latest/setup_script.html#cx-freeze-executable
https://cx-freeze.readthedocs.io/en/latest/script.html#cmdoption-manifest
